core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1217,26 +1217,41 @@ public String toString() {
      * @return true if the remote segment's epoch/offsets are within the leader epoch lineage of the partition.
      */
     // Visible for testing
-    public static boolean isRemoteSegmentWithinLeaderEpochs(RemoteLogSegmentMetadata segmentMetadata,
-                                                            long logEndOffset,
-                                                            NavigableMap<Integer, Long> leaderEpochs) {
+    static boolean isRemoteSegmentWithinLeaderEpochs(RemoteLogSegmentMetadata segmentMetadata,
+                                                     long logEndOffset,
+                                                     NavigableMap<Integer, Long> leaderEpochs) {
         long segmentEndOffset = segmentMetadata.endOffset();
         // Filter epochs that does not have any messages/records associated with them.
         NavigableMap<Integer, Long> segmentLeaderEpochs = buildFilteredLeaderEpochMap(segmentMetadata.segmentLeaderEpochs());
         // Check for out of bound epochs between segment epochs and current leader epochs.
-        Integer segmentFirstEpoch = segmentLeaderEpochs.firstKey();
         Integer segmentLastEpoch = segmentLeaderEpochs.lastKey();
-        if (segmentFirstEpoch < leaderEpochs.firstKey() || segmentLastEpoch > leaderEpochs.lastKey()) {
+        if (segmentLastEpoch < leaderEpochs.firstKey() || segmentLastEpoch > leaderEpochs.lastKey()) {
+            LOGGER.debug("Segment {} is not within the partition leader epoch lineage. " +
+                            "Remote segment epochs: {} and partition leader epochs: {}",
+                    segmentMetadata.remoteLogSegmentId(), segmentLeaderEpochs, leaderEpochs);
+            return false;
+        }
+        // There can be overlapping remote log segments in the remote storage. (eg)
+        // leader-epoch-file-cache: {(5, 10), (7, 15), (9, 100)}
+        // segment1: offset-range = 5-50, Broker = 0, epochs = {(5, 10), (7, 15)}
+        // segment2: offset-range = 14-150, Broker = 1, epochs = {(5, 14), (7, 15), (9, 100)}, after leader-election.
```

Comment:
Could I understand in which case, we'll upload 2 segments with offset overlapped? I thought the `findHighestRemoteOffset` should help get the `copiedOffset` for the new leader, no?

Reply:
yes, you're right. The `findHighestRemoteOffset` will return the `copiedOffset` for the new leader. 

The case mentioned can happen as the logic to rotate the segments is independent of brokers in the cluster. 

Broker-0 can have local-log segments like:
   segment0 = 0-50
   segment1 = 51-100
   active-segment = 101-120
 
 and it uploaded both segment0 and segment1 to remote storage.
 
 Broker1 can have local-log segments like:
   segment0 = 0 - 80
   active-segment = 81-120
   
When Broker1 becomes leader, it findHighestRemoteOffset as 100, then it has to upload the current active segment segment once it gets rotated.

Reply:
OK, the Broker1 having segment size different from broker0 could be because the Broker1 has the larger `log.segment.bytes` than Broker0. That makes sense. It's quite rare to see, though.

Reply:
Given that we roll the log based on `log.roll.ms` and the server startup time can vary between the replicas, this is quite common.